code/datums/elements/immerse.dm

```diff
@@ -0,0 +1,387 @@
+	SIGNAL_HANDLER
+	if(source.movement_type & (FLYING|FLOATING))
+		return
+	add_immerse_overlay(source)
+	for(var/mob/living/buckled_mob as anything in source.buckled_mobs)
+		add_immerse_overlay(buckled_mob)
+
+/**
+ * Called when a movable exits the turf. If its new location is not in the list of turfs with this element,
+ * Remove the movable from the element.
+ */
+/datum/element/immerse/proc/on_atom_exited(turf/source, atom/movable/exited, direction)
+	SIGNAL_HANDLER
+	if(!(exited.loc in attached_turfs_and_movables))
+		remove_from_element(source, exited)
+	else
+		LAZYREMOVE(attached_turfs_and_movables[source], exited)
+		LAZYADD(attached_turfs_and_movables[exited.loc], exited)
+
+///Remove any signal, overlay, trait given to the movable and reference to it within the element.
+/datum/element/immerse/proc/remove_from_element(turf/source, atom/movable/movable)
+	var/atom/movable/buckled
+	if(isliving(movable))
+		var/mob/living/living_mob = movable
+		buckled = living_mob.buckled
+	try_unimmerse(movable, buckled)
+
+	UnregisterSignal(movable, list(COMSIG_LIVING_SET_BUCKLED, COMSIG_QDELETING))
+	REMOVE_TRAIT(movable, TRAIT_IMMERSED, ELEMENT_TRAIT(src))
+	LAZYREMOVE(attached_turfs_and_movables[source], movable)
+
+/// A band-aid to keep the (unique) visual overlay from scaling and rotating along with its owner. I'm sorry.
+/datum/element/immerse/proc/on_update_transform(mob/living/source, resize, new_lying_angle, is_opposite_angle)
+	SIGNAL_HANDLER
+	var/matrix/new_transform = matrix()
+	new_transform.Scale(1/source.current_size)
+	new_transform.Turn(-new_lying_angle)
+
+	var/atom/movable/immerse_overlay/vis_overlay = immersed_movables[source]
+	if(is_opposite_angle)
+		vis_overlay.transform = new_transform
+		vis_overlay.adjust_living_overlay_offset(source)
+		return
+
+	/**
+	 * Here, we temporarily switch from the offset of the mutable appearance to one for movable used as visual overlay.
+	 * Why? While visual overlays can be animated, their fixed point stays at the center of the icon of the atom
+	 * they're attached to and not theirs, which can make manipulating the transform var a pain, but because
+	 * we cannot do that with normal overlay or filters (reliably), we have to bend a knee and try to compensate it.
+	 */
+	vis_overlay.overlays = list(vis_overlay.overlay_appearance)
+
+	/// Oh, yeah, didn't I mention turning a visual overlay affects its pixel x/y/w/z too? Yeah, it sucks.
+	var/new_x = vis_overlay.extra_width
+	var/new_y = vis_overlay.extra_height
+	var/old_div = source.current_size/resize
+	var/offset_lying = source.rotate_on_lying ? PIXEL_Y_OFFSET_LYING : source.get_pixel_y_offset_standing(source.current_size/resize)
+	switch(source.lying_prev)
+		if(270)
+			vis_overlay.pixel_x += -offset_lying / old_div
+		if(90)
+			vis_overlay.pixel_x += offset_lying / old_div
+		if(0)
+			vis_overlay.pixel_y += -source.get_pixel_y_offset_standing(source.current_size/resize) / old_div
```

Comment:
do you want pixel_y or pixel_z here?

Reply:
It makes no difference... wait, are you implying I'm accidentally using both pixel y and z vars?